### PR TITLE
Viite-3103 reset split links when terminating or reverting project links

### PR DESCRIPTION
**Pull request by jounijoh**
Added function identifyAndHandleSplitProjectLinks to 
identify and reset unnecessary split links when terminating or reverting project links to original values.
- identifySplitsToHandle groups project links by linear location id and sorts the groups by id
- After sorting, the higher id project links are validated for removing by checking if it has same address m values as the project link with the same linear location id 
- After validation the newer links are removed and the original link is saved with its geometry updated and its connected link id removed 

batchUpdateProjectLinksToReset now updates all terminated project links with same linear location to the values of the original road address data so the splits can be identified.

Added methods updateProjectLinkGeometry and handleSplitLinkRemovalAndUpdate to handle the database updates

### CODE REVIEW

File: viite-backend/viite-main/src/main/scala/fi/liikennevirasto/viite/ProjectService.scala [hunk 1]
```diff
@@ -1623,6 +1625,63 @@ class ProjectService(
...
+    val kgvRoadLinks = roadLinkService.getCurrentAndComplementaryRoadLinks(Set(originalLink.linkId))
+    kgvRoadLinks.find(_.linkId == originalLink.linkId).foreach { kgvGeometry =>
+      val updatedGeom = GeometryUtils.truncateGeometry3D(kgvGeometry.geometry, originalLink.startMValue, originalLink.endMValue)
+      projectLinkDAO.updateProjectLinkGeometry(originalLink.id, updatedGeom)
+      val updatedLink = originalLink.copy(geometry = updatedGeom)
+      updatedProjectLinks += updatedLink
+    }
+  }
+
+  // If both project links have same the same address m values after termination, the other can be removed.
```
Comment (marjukka-kokkonen-sitowise) [inline viite-backend/viite-main/src/main/scala/fi/liikennevirasto/viite/ProjectService.scala:1680]: redundancy in the text: "same the same"

File: viite-backend/viite-main/src/main/scala/fi/liikennevirasto/viite/ProjectService.scala [hunk 2]
```diff
@@ -1216,14 +1213,12 @@ class ProjectService(
         case _ =>
       })
     revertRoadName(projectId, roadPart.roadNumber)
-    if (recalculate)
```
Comment (jounijoh) [inline viite-backend/viite-main/src/main/scala/fi/liikennevirasto/viite/ProjectService.scala:1219]: Never in use

File: viite-backend/viite-main/src/test/scala/fi/liikennevirasto/viite/dao/ProjectLinkDAOSpec.scala [hunk 3]
```diff
@@ -453,7 +453,7 @@ class ProjectLinkDAOSpec extends FunSuite with Matchers {
         link.endAddrMValue shouldBe originalEndAddrMValue
         link.startMValue shouldBe originalStartMValue
         link.endMValue shouldBe originalEndMValue
-        link.status shouldBe RoadAddressChangeType.NotHandled
+        link.status shouldBe RoadAddressChangeType.Termination
```
Comment (LauriLitmanenSW) [inline viite-backend/viite-main/src/test/scala/fi/liikennevirasto/viite/dao/ProjectLinkDAOSpec.scala:456]: Jos kaikki linkit palautetaan alkuperäiseen tilaan niin eikö niiden kuuluisi olla NotHandled vai missasinko jotain?

File: viite-backend/viite-main/src/main/scala/fi/liikennevirasto/viite/dao/ProjectLinkDAO.scala [hunk 4]
```diff
@@ -751,17 +755,19 @@ println(sql)
           updatePS.setInt(5, pl.administrativeClass.value)
           updatePS.setLong(6, pl.startAddrMValue)
           updatePS.setLong(7, pl.endAddrMValue)
-          updatePS.setInt(8, pl.calibrationPointTypes._1.value)
-          updatePS.setInt(9, pl.calibrationPointTypes._2.value)
+          updatePS.setLong(8, pl.startAddrMValue)
```
Comment (LauriLitmanenSW) [inline viite-backend/viite-main/src/main/scala/fi/liikennevirasto/viite/dao/ProjectLinkDAO.scala:758]: originalStartAddrMValue?

File: viite-backend/viite-main/src/main/scala/fi/liikennevirasto/viite/dao/ProjectLinkDAO.scala [hunk 5]
```diff
@@ -751,17 +755,19 @@ println(sql)
...
-          updatePS.setInt(16, RoadAddressChangeType.NotHandled.value) // as it's the original value before changes
-          updatePS.setLong(17, pl.linearLocationId)
-          updatePS.setLong(18, pl.id)
+          updatePS.setInt(12, pl.calibrationPointTypes._1.value)
+          updatePS.setInt(13, pl.calibrationPointTypes._2.value)
+          updatePS.setInt(14, pl.sideCode.value)
+          updatePS.setLong(15, pl.ely)
+          updatePS.setDouble(16, pl.startMValue)
+          updatePS.setDouble(17, pl.endMValue)
+          updatePS.setInt(18, RoadAddressChangeType.Termination.value)
```
Comment (LauriLitmanenSW) [inline viite-backend/viite-main/src/main/scala/fi/liikennevirasto/viite/dao/ProjectLinkDAO.scala:768]: Miksi Termination?

**Review by LauriLitmanenSW — COMMENTED** [review]
Pari nillitystä ja kyssäriä

File: viite-backend/viite-main/src/test/scala/fi/liikennevirasto/viite/dao/ProjectLinkDAOSpec.scala (hunk 3 above)
Comment (jounijoh) [reply viite-backend/viite-main/src/test/scala/fi/liikennevirasto/viite/dao/ProjectLinkDAOSpec.scala:456]: Käytössä vain lakkautuksissa, korjasin testikuvauksen

**Pull request merged**